compiler/damlc/lib/DA/Cli/Damlc.hs

```diff
@@ -537,39 +530,144 @@ createProjectPackageDb (AllowDifferentSdkVersions allowDiffSdkVersions) lfVersio
                           [".daml", ".hie", ".hi"]
                     ]
-            forM_ dalfs $ \dalf -> do
-                let path = dbPath </> ZipArchive.eRelativePath dalf
-                createDirectoryIfMissing True (takeDirectory path)
-                BSL.writeFile path (ZipArchive.fromEntry dalf)
-            forM_ confFiles $ \conf ->
-                BSL.writeFile
-                    (dbPath </> "package.conf.d" </>
-                     (takeFileName $ ZipArchive.eRelativePath conf))
-                    (ZipArchive.fromEntry conf)
-            forM_ srcs $ \src -> do
-                let path = dbPath </> ZipArchive.eRelativePath src
-                write path (ZipArchive.fromEntry src)
-            pure sdkVersion
+            pure ([sdkVersion], confFiles, dalfs, srcs)
     let uniqSdkVersions = nubSort sdkVersions
-    -- if there are no package dependencies, sdkVersions will be empty
-    unless (length uniqSdkVersions <= 1 || allowDiffSdkVersions) $
-        fail $
-        "Package dependencies from different SDK versions: " ++
-        intercalate ", " uniqSdkVersions
-    ghcPkgPath <-
-        if isWindows
-          then locateRunfiles "rules_haskell_ghc_windows_amd64/bin"
-          else locateRunfiles "ghc_nix/lib/ghc-8.6.5/bin"
-    callProcess
-        (ghcPkgPath </> exe "ghc-pkg")
-        [ "recache"
-        -- ghc-pkg insists on using a global package db and will try
-        -- to find one automatically if we don’t specify it here.
-        , "--global-package-db=" ++ (dbPath </> "package.conf.d")
-        , "--expand-pkgroot"
-        ]
+    if | length uniqSdkVersions <= 1 ->
+           do forM_ dalfs $ \dalf -> do
+                  let path = dbPath </> ZipArchive.eRelativePath dalf
+                  createDirectoryIfMissing True (takeDirectory path)
+                  BSL.writeFile path (ZipArchive.fromEntry dalf)
+              forM_ confFiles $ \conf ->
+                  BSL.writeFile
+                      (dbPath </> "package.conf.d" </>
+                       (takeFileName $ ZipArchive.eRelativePath conf))
+                      (ZipArchive.fromEntry conf)
+              forM_ srcs $ \src -> do
+                  let path = dbPath </> ZipArchive.eRelativePath src
+                  write path (ZipArchive.fromEntry src)
+              ghcPkgPath <- getGhcPkgPath
+              callProcess
+                  (ghcPkgPath </> exe "ghc-pkg")
+                  [ "recache"
+                  -- ghc-pkg insists on using a global package db and will try
+                  -- to find one automatically if we don’t specify it here.
+                  , "--global-package-db=" ++ (dbPath </> "package.conf.d")
+                  , "--expand-pkgroot"
+                  ]
+       | length uniqSdkVersions > 1 && optAllowDifferentSdks opts ->
+            -- when we compile packages with different sdk versions, we need to generate the interface files
```

Comment:
Would it be better to just have one case to compiling the package db, whether it be with one or multiple SDKs? I guess it would be slower...

Reply:
That would be a lot slower.

Reply:
currently that would just fail. but yes, as soon as you have cross sdk dependencies, we generate the interface files.